corebehrt/functional/utils.py

```diff
@@ -106,9 +144,91 @@ def remove_missing_timestamps(df: pd.DataFrame) -> pd.DataFrame:
     return df[df.TIMESTAMP.notna()]
 
 
-def get_first_event_by_pid(df: pd.DataFrame):
+def get_first_event_by_pid(df: pd.DataFrame, timestamp_column: str) -> pd.DataFrame:
     """
-    Assumes that the table has a column named PID and TIMESTAMP.
+    Assumes that the table has a column named PID and abspos.
     Get the first event for each PID in the table.
     """
-    return df.groupby("PID").TIMESTAMP.min()
+    return df.groupby("PID")[timestamp_column].min()
+
+
+def select_random_subset(data: dd.DataFrame, n: int) -> dd.DataFrame:
+    """
+    Assumes that the table has a column named PID.
+    Returns a new table with a random subset of n PIDs.
+    """
+    if n >= len(data):
+        return data
+    pids = data["PID"].unique().compute().tolist()
+    random.seed(42)
+    random.shuffle(pids)
+    pids = pids[:n]
+    return filter_table_by_pids(data, pids)
+
+
+def truncate_patient(
+    patient: pd.DataFrame, background_length: int, max_len: int, sep_token: str
+) -> pd.DataFrame:
+    """
+    Assumes patient is pd.DataFrame.
+    Truncate patient to max_len, keeping background.
+    """
+    if len(patient["concept"]) <= max_len:
+        return patient
+
+    truncation_length = max_len - background_length
+    if patient["concept"].iloc[-truncation_length] == sep_token:
+        truncation_length -= 1
+
+    truncated_patient = pd.concat(
+        [patient.iloc[:background_length], patient.iloc[-truncation_length:]],
+        ignore_index=True,
+    )
+
+    return truncated_patient
+
+
+def truncate_data(
+    data: dd.DataFrame,
+    max_len: int,
+    vocabulary: dict,
+    truncate_function: Callable = truncate_patient,
+) -> dd.DataFrame:
+    """
+    Assumes table has a column named PID.
+    Truncate the data to max_len. CLS and SEP tokens are kept if present.
+    Uses truncate_patient as default truncate function.
+    """
+    background_length = get_background_length_dd(data, vocabulary)
+
+    truncated_data = (
+        data.groupby("PID")[list(data.columns)]
+        .apply(
+            lambda x: truncate_function(
+                x, background_length, max_len, vocabulary.get("[SEP]")
+            ),
+            meta={col: dtype for col, dtype in data.dtypes.items()},
+        )
+        .reset_index(drop=True)
+    )
+
+    return truncated_data
+
+
+def get_gender_token(gender: str, vocabulary: dict) -> int:
+    """
+    Retrieves the token for the specified gender from the vocabulary.
+    Assumes that the gender starts with BG_GENDER_.
+    """
+    gender_key = f"BG_GENDER_{gender}"
+    try:
+        return vocabulary[gender_key]
+    except KeyError:
```

Comment:
Is this not the default behavior of a dictionary, except slightly more detailed? I suggest just letting the dictionary throw its standard error for clarification

Reply:
Ok